File: 75.sort-colors.js
/*
 * @lc app=leetcode id=75 lang=javascript
 *
 * [75] Sort Colors
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function (nums) {
  let p0 = 0;
  let p2 = nums.length - 1;
  for (let i = 0; i <= p2; ++i) {
    if (nums[i] === 0) {
      [nums[i], nums[p0]] = [nums[p0], nums[i]];
      p0++;
    } else if (nums[i] === 2) {
      [nums[i], nums[p2]] = [nums[p2], nums[i]];
      p2--;
      i--;
    }
  }
};
// @lc code=end

File: solutions/75.sort-colors.rb
# @param {Integer[]} nums
# @return {Void} Do not return anything, modify nums in-place instead.
def sort_colors(nums)
  p = 0
  (0...nums.length).each do |i|
    if nums[i] == 0
      swap(nums, i, p)
      p += 1
    end
  end
  (p...nums.length).each do |i|
    if nums[i] == 1
      swap(nums, i, p)
      p += 1
    end
  end
end

def swap(nums, i, j)
  nums[i], nums[j] = nums[j], nums[i]
end
